sdk/identity/azure-identity/src/main/java/com/azure/identity/ChainedTokenCredential.java

```diff
@@ -38,27 +40,31 @@
     public Mono<AccessToken> getToken(TokenRequestContext request) {
         List<CredentialUnavailableException> exceptions = new ArrayList<>(4);
         return Flux.fromIterable(credentials)
-                   .flatMap(p -> p.getToken(request).onErrorResume(CredentialUnavailableException.class, t -> {
-                       exceptions.add(t);
-                       return Mono.empty();
-                   }), 1)
-                   .next()
-                   .switchIfEmpty(Mono.defer(() -> {
+            .flatMap(p -> p.getToken(request)
+                .doOnNext(t -> logger.info("Azure Identity => Attempted credential {} returns a token",
+                    p.getClass().getSimpleName()))
+                .onErrorResume(CredentialUnavailableException.class, t -> {
+                    exceptions.add(t);
+                    logger.info("Azure Identity => Attempted credential {} is unavailable.",
+                        p.getClass().getSimpleName());
+                    return Mono.empty();
+                }), 1)
+            .next()
+            .switchIfEmpty(Mono.defer(() -> {
+                StringBuilder message = new StringBuilder("Tried "
+                    + credentials.stream().map(c -> c.getClass().getSimpleName())
+                    .collect(Collectors.joining(", "))
+                    + " but failed to acquire a token for any of them. Please verify the"
+                    + " environment for the credentials"
+                    + " and see more details in the causes below.");
```

Comment:
Will it be clear where the causes are? What identifies the causes if they want to find them quickly in the log?

Reply:
This error message has been improved through this PR: https://github.com/Azure/azure-sdk-for-java/pull/9022